lib/wibox/layout/overflow.lua

```diff
@@ -0,0 +1,408 @@
+    if is_y then
+        height = math.huge
+    else
+        width = math.huge
+    end
+
+    -- First, determine widget sizes.
+    -- Only when the content doesn't fit and needs scrolling should
+    -- we reduce content size to make space for a scrollbar.
+    for _, widget in pairs(widgets) do
+        local w, h = base.fit_widget(self, context, widget, width, height)
+
+        if is_y then
+            used_max = math.max(used_max, w)
+            used_in_dir = used_in_dir + h
+        else
+            used_in_dir = used_in_dir + w
+            used_max = math.max(used_max, h)
+        end
+    end
+
+    local spacing = self._private.spacing * (num_widgets - 1)
+    used_in_dir = used_in_dir + spacing
+
+    local need_scrollbar = used_in_dir > avail_in_dir and scrollbar_enabled
+
+    -- If the direction perpendicular to scrolling (e.g. width in vertical
+    -- scrolling) is not fully covered by any of the widgets, we can add our
+    -- scrollbar width to that value. Otherwise widget size will be reduced
+    -- during `layout` to make space for the scrollbar.
+    if need_scrollbar
+        and (
+            (is_y and used_max < orig_width)
+            or (not is_y and used_max < orig_height)
+        ) then
+        used_max = used_max + scrollbar_width
+    end
+
+    if is_y then
+        return used_max, used_in_dir
+    else
+        return used_in_dir, used_max
+    end
+end
+
+-- Adapted version of `wibox.layout.fixed.layout`
+function overflow:layout(context, orig_width, orig_height)
+    local result = {}
+    local is_y = self._private.dir == "y"
+    local widgets = self._private.widgets
+    local avail_in_dir = is_y and orig_height or orig_width
+    local scrollbar_width = self._private.scrollbar_width
+    local scrollbar_enabled = self._private.scrollbar_enabled
+    local scrollbar_position = self._private.scrollbar_position
+    local width, height = orig_width, orig_height
+    local widget_x, widget_y = 0, 0
+    local used_in_dir, used_max = 0, 0
+
+    -- Set the direction covered by scrolling to the maximum value
+    -- to allow widgets to take as much space as they want.
+    if is_y then
+        height = math.huge
+    else
+        width = math.huge
+    end
+
+    -- First, determine widget sizes.
+    -- Only when the content doesn't fit and needs scrolling should
+    -- we reduce content size to make space for a scrollbar.
+    for _, widget in pairs(widgets) do
+        local w, h = base.fit_widget(self, context, widget, width, height)
+
+        if is_y then
+            used_max = math.max(used_max, w)
+            used_in_dir = used_in_dir + h
+        else
+            used_in_dir = used_in_dir + w
+            used_max = math.max(used_max, h)
+        end
+    end
+
+    used_in_dir = used_in_dir + self._private.spacing * (#widgets-1)
+
+    -- Save size for scrolling behavior
+    self._private.avail_in_dir = avail_in_dir
+    self._private.used_in_dir = used_in_dir
```

Comment:
What if the overflow container is shown in more than one place with different sizes? Assigning size-information to "something in `self`" sounds like either a bug or a bug-waiting-to-happen to me.

This problem is why the scroll container does all that funky business with weak tables.

(Being able to show a widget in more than one place is great for users, but it brings so many complications...)

Reply:
This might be a showstopper then. Knowing the exact geometry is pretty much required for (useful) scrolling to work.

I did check out the `scroll` container, but can't claim to understand its funky business yet. And even if I did, combining funky business with user input sounds like a "bug-waiting-to-happen" as well.

It would probably require changes to the core of the widget system so that when something attempts to re-use an existing instance, that widget may request to be cloned instead (maybe with custom logic attached to the clone function).

Reply:
could it be imposed some limitation against multiple instances? (so it will throw a warning and won't attempt to draw more than one instance)

Reply:
That would certainly be an option and I would be fine with imposing such a limitation.
But this would certainly be the first widget to do so, although looking at the rework of `wibox.widget.graph` it's not the only one that would benefit from adding widgets with this kind of limitation.

Reply:
arent wibox/wibar/popup also could have only one instance of a given geometry?

Reply:
> could it be imposed some limitation against multiple instances? 

Not really. At least I wouldn't know how.

The systray widget has the same limitation. The current code basically works on hopes and dreams: Wherever the systray widget was drawn last is where it is visible.

There is some special magic in `wibox.drawable` in the (undocumented on purpose) method `_set_systray_widget`. This calls `wibox.hierarchy.count_widget` which makes the hierarchy count the number of systray widgets in hierarchies. When re-computing the layout, the count is checked before/after and if it goes to zero, a special (also undocumented) C function is called to really hide the systray.

So, a lot of work just to make sure hiding the systray widget really hides it. This is not directly related to this issue, but should just show that I already had lots of problems making the systray work.

Reply:
wait, but for systray we have `container.only_on_screen()`

i guess, that shouldnt be much harder to implement `container.display_only_once` or so?

Reply:
That container works via the context argument that is passed in to widgets. When `context.screen` is the expected screen, it works normally. Otherwise, `:fit` returns `0, 0` and `:layout` does nothing. Anyway, you can still show a widget multiple times on a given screen.

(And I guess this widget does not work correctly when used in a titlebar and you move the client to another screen. Or when used in a wibox and that wibox is moved to another screen. Nothing invalidates the caches or triggers a relayout in these cases.)

Reply:
i think we could create the new container to accomodate this case, by just storing which parent widget first displayed it, and ignore all other fit/layout request from different parent widgets?

Reply:
however, from the second thought, i feel like it should be acomodated here in overflow container - just have internal table for storing data for multiple parent widgets

Reply:
like

```lua
    self._private.instances[parent_id].avail_in_dir = avail_in_dir
    self._private.instances[parent_id].used_in_dir = used_in_dir
```